cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityApplier.java

```diff
@@ -273,12 +284,34 @@ private void deployApp(SystemApplication application) throws Exception {
                  application.getArtifact(), configString, NOOP_PROGRAM_TERMINATOR, null, null);
   }
 
-  private boolean isAppDeployed(ApplicationId applicationId) throws Exception {
+  // Returns true if capability applier should try to deploy this application. 2 conditions when it returns true:
+  // 1. Either the application is not deployed before.
+  // 2. If application is deployed before then the app artifact of the deployed application is not the latest one
+  //    available.
+  private boolean shouldDeployApp(ApplicationId applicationId, SystemApplication application) throws Exception {
     try {
-      applicationLifecycleService.getAppDetail(applicationId);
-      return true;
+      ApplicationDetail currAppDetail = applicationLifecycleService.getAppDetail(applicationId);
+
+      ArtifactSummary summary = application.getArtifact();
+      NamespaceId artifactNamespace =
+        ArtifactScope.SYSTEM.equals(summary.getScope()) ? NamespaceId.SYSTEM : applicationId.getParent();
+      ArtifactRange range = new ArtifactRange(artifactNamespace.getNamespace(), summary.getName(),
+        ArtifactVersionRange.parse(summary.getVersion()));
+      // this method will not throw ArtifactNotFoundException, if no artifacts in the range, we are expecting an empty
+      // collection returned.
+      List<ArtifactDetail> artifactDetail = artifactRepository.getArtifactDetails(range, 1, ArtifactSortOrder.DESC);
+      if (artifactDetail.isEmpty()) {
+        throw new ArtifactNotFoundException(range.getNamespace(), range.getName());
+      }
+
+      ArtifactDetail latestArtifact = artifactDetail.iterator().next();
+      // Compare the version of app artifact for deployed application and the latest available version of that
+      // same artifact. If same means no need to deploy the application again.
+      return !currAppDetail.getArtifact().getVersion().equals(
```

Comment:
What will happen in following scenario:
1. Capability deploys the app with latest version
2. Someone deletes the latest artifact
3. This will deploy the app with older artifact, is that expected?

Reply:
Yes that is exactly what would happen.

Reply:
It is better to add a log message if that will happen, otherwise it is confusing since app behavior will change

Reply:
Added the log below just before deploying application at which place I feel it would solve this purpose and also add log for normal scenario where we are deploying capability.